File: content/snippets/scala2/interview_lcci/02/03_delete_middle_node.scala
object Lcci0203DeleteMiddleNode {
  final case class ListNode(var value: Int, var next: ListNode | Null = null)

  def buildList(nums: Vector[Int]): ListNode | Null = {
    val dummy = ListNode(0, null)
    var tail: ListNode = dummy
    nums.foreach { n =>
      tail.next = ListNode(n, null)
      tail = tail.next.asInstanceOf[ListNode]
    }
    dummy.next
  }

  def deleteMiddleNode(node: ListNode | Null): Boolean = {
    if (node == null || node.asInstanceOf[ListNode].next == null) {
      return false
    }
    val cur = node.asInstanceOf[ListNode]
    val next = cur.next.asInstanceOf[ListNode]
    cur.value = next.value
    cur.next = next.next
    true
  }

  def printList(head: ListNode | Null): Unit = {
    val out = scala.collection.mutable.ArrayBuffer.empty[Int]
    var cur = head
    while (cur != null) {
      val node = cur.asInstanceOf[ListNode]
      out += node.value
      cur = node.next
    }
    println(out.mkString(" "))
  }

  def main(args: Array[String]): Unit = {
    val head = buildList(Vector(1, 2, 3, 4, 5)).asInstanceOf[ListNode]
    deleteMiddleNode(head.next.asInstanceOf[ListNode].next)
    printList(head)
  }
}

File: content/snippets/typescript/interview_lcci/02/03_delete_middle_node.ts
class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function buildList(nums: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let tail = dummy;
  for (const n of nums) {
    tail.next = new ListNode(n);
    tail = tail.next;
  }
  return dummy.next;
}

function deleteMiddleNode(node: ListNode | null): boolean {
  if (node === null || node.next === null) {
    return false;
  }
  node.val = node.next.val;
  node.next = node.next.next;
  return true;
}

function printList(head: ListNode | null): void {
  const out: number[] = [];
  while (head !== null) {
    out.push(head.val);
    head = head.next;
  }
  console.log(out.join(" "));
}

function main(): void {
  const head = buildList([1, 2, 3, 4, 5])!;
  deleteMiddleNode(head.next!.next!);
  printList(head);
}

main();
